cmd/root.go

```diff
@@ -91,52 +90,131 @@ func NewRootCommand() *cobra.Command {
 	return cmd
 }
 
-// initViperConfigFile reads in config file and ENV variables if set.
-func initViperConfigFile() {
+func initViperAndProfile() {
 	l := logger.Get()
 
+	// If no configuration file location is specified, use the default configuration file location
 	if cfgFile == "" {
-		l.Debug().Msgf("No configuration file specified. Determining default configuration file location: $HOME/.pingctl/config.yaml")
-		cfgFile = defaultCfgFile
-		l.Debug().Msgf("Determined default configuration file location: %s", cfgFile)
-
-		// Make sure the default config file exists, and if not, seed a new file
-		if _, err := os.Stat(cfgFile); os.IsNotExist(err) {
-			l.Debug().Msgf("Default configuration file does not exists. Seeding a new file at location: %s", cfgFile)
-
-			// MkdirAll does nothing if directories already exist. Create needed directories for config file location.
-			err := os.MkdirAll(filepath.Dir(cfgFile), os.ModePerm)
-			if err != nil {
-				l.Fatal().Err(err).Msgf("Failed to make directories needed for filepath: %s", cfgFile)
-			}
+		initDefaultConfigFile()
+	}
 
-			// SafeWriteConfigAs writes current configuration to a given filename if it does not exist.
-			err = viper.SafeWriteConfigAs(cfgFile)
-			if err != nil {
-				l.Fatal().Err(err).Msgf("Failed to create configuration file at: %s", cfgFile)
-			}
-		}
+	//Configure the main viper instance
+	initMainViper()
+
+	// Prefer parameter, then environment variable, then configuration file
+	// like with viper hierarchy. Finally default to @defaultProfileName if not found
+	// NOTE: this is needed because parameter and env var are not bound to
+	// the main viper instance
+	if profileName == "" {
+		profileName = os.Getenv(profiles.ProfileOption.EnvVar)
+	}
+	if profileName == "" {
+		mainViper := profiles.GetMainViper()
+		profileName = mainViper.GetString(profiles.ProfileOption.ViperKey)
+	}
+	if profileName == "" {
+		profileName = defaultProfileName
+	}
+
+	l.Debug().Msgf("Using configuration profile: %s", profileName)
+
+	// Configure the profile viper instance
+	if err := profiles.SetProfileViperWithProfile(profileName); err != nil {
+		output.Print(output.Opts{
```

Comment:
Should this return/exit right away if an error was returned? This also applies to other output.Print calls that deal with fatal errors.

Reply:
Good question. All 7 FatalMessage outputs that exit right away are in root.go and deal with config file initialization and viper initialization. I think it makes sense to exit early if those processes error as the tool could run unpredictably if the configuration is not correct.